File: Games/Alex/src/de/hsh/alexander/level/PacManLevel.java
package de.hsh.alexander.level;

import common.actor.*;
import common.config.WindowConfig;
import common.util.Logger;
import de.hsh.alexander.actor.collectables.DataCoin;
import de.hsh.alexander.actor.collectables.Invisible;
import de.hsh.alexander.actor.level_elements.Condensator;
import de.hsh.alexander.actor.level_elements.SMD;
import de.hsh.alexander.actor.player.PacMan;
import de.hsh.alexander.actor.player.PacMan1;
import de.hsh.alexander.actor.player.PacMan2;
import javafx.beans.property.SimpleIntegerProperty;
import javafx.scene.canvas.Canvas;
import javafx.scene.input.KeyEvent;
import javafx.scene.input.MouseEvent;

import java.util.Observable;

import static de.hsh.alexander.PacManGame.gameFinishedMessage;

abstract public class PacManLevel extends Level {

    static final int background_width  = 950;
    static final int background_height = 950;


    PacManLevel( Canvas gameCanvas ) {
        super( gameCanvas );
    }

    @Override
    public void reset( Canvas gameCanvas ) {
        super.reset( gameCanvas );
        gameCanvas.setOnMouseClicked( this::onMouseClick );
        //Logger.log( this.getClass() + ": Resettet Level" );
        isGameFinished();
    }

    private void onMouseClick( final MouseEvent clickEvent ) {
        final double x = clickEvent.getX();
        final double y = clickEvent.getY();
        //Logger.log( this.getClass() + ": Clicked at : (" + x + ", " + y + ")" );
        this.addCollectable(
                new DataCoin( x, y )
                           );

    }

    public SimpleIntegerProperty getPacMan1Property() {
        return getPlayers().stream()
                           .filter( player -> player instanceof PacMan1 )
                           .map( player -> (PacMan1) player )
                           .map( PacMan::getPointProperty ).findFirst().get();
    }

    public SimpleIntegerProperty getPacMan2Property() {
        return getPlayers().stream()
                           .filter( player -> player instanceof PacMan2 )
                           .map( player -> (PacMan) player )
                           .map( PacMan::getPointProperty ).findFirst().get();
    }

    @Override
    public void keyboardInput( final KeyEvent keyEvent ) {
        getPlayers().forEach( pacMan -> pacMan.move( keyEvent ) );
    }

    @Override
    public void update( Observable o, final Object arg ) {
        if ( o instanceof Collectable ) {
            final Collectable c = (Collectable) o;
            if ( arg instanceof String ) {
                if ( arg.equals( Collectable.collected ) ) {
                    coinCollected( c );
                }
            }
            if ( c instanceof Invisible ) {
                Invisible i = (Invisible) c;
                Logger.log( "PacManLevel : invisible collected : " + i );
            }
        }
        else {
            Logger.log( this.getClass() + ": unknown observable=" + o + " , arg=" + arg );
        }
    }

    private void coinCollected( final Collectable c ) {
        if ( c instanceof DataCoin ) {
            final Actor a = c.getCollector();
            if ( a instanceof PacMan ) {
                final PacMan p = (PacMan) a;
                p.addPoint();
            }
        }
        this.collected( c );
    }

    protected boolean isGameFinished() {
        if ( this.getCollectables().size() == 0 ) {
            this.setChanged();
            this.notifyObservers( gameFinishedMessage );
            return true;
        }
        return false;
    }

    void createDataCoins( Canvas gameCanvas ) {
        for ( int y = 0 ; y < WindowConfig.window_height ; y += 50 ) {
            for ( int x = 0 ; x < WindowConfig.window_width ; x += 50 ) {
                final DataCoin  d  = new DataCoin( x, y );
                if (CollisionCheck.isInBounds( d, gameCanvas )) {
                    addCollectable( d );
                }
            }
        }
    }

    void addEasterEgg( Canvas gameCanvas, final int x, final int y ) {
        addCollectable( new Invisible( x, y ) );
    }

    void addPlayers( final int pacMan1_x, final int pacMan1_y, final int pacMan2_x, final int pacMan2_y ) {
        addPlayer( new PacMan1( pacMan1_x, pacMan1_y ) );
        addPlayer( new PacMan2( pacMan2_x, pacMan2_y ) );
    }

    void fillPins( Canvas gameCanvas ) {

        addLevelElement( gameCanvas, new Condensator( 200, 600, 1 ) );

        for ( int i = 350 ; i < 500 ; i += 40 ) {
            addLevelElement( gameCanvas, new Condensator( 700, i, 0 ) );
        }


        final int  end       = 580;
        final byte increment = 19;

        addCondensators_y( gameCanvas, 285, end, increment, 525 );
        addCondensators_x( gameCanvas, 545, 805, increment, 705 );


        addLevelElement( gameCanvas, new Condensator( 525 + increment, end + increment ) );
    }


    private void addCondensators_y( Canvas gameCanvas,
                                    final int start,
                                    final int end,
                                    final int increment,
                                    final int x ) {
        for ( int y = start ; y < end ; y += increment ) {
            addLevelElement( gameCanvas, new Condensator( x, y ) );
            addLevelElement( gameCanvas, new Condensator( x + increment, y ) );
        }
    }

    private void addCondensators_x( Canvas gameCanvas,
                                    final int start,
                                    final int end,
                                    final int increment,
                                    final int y ) {
        for ( int x = start ; x < end ; x += increment ) {
            addLevelElement( gameCanvas, new Condensator( x, y ) );
            addLevelElement( gameCanvas, new Condensator( x + increment, y ) );
        }
    }


    void addSMDs( Canvas gameCanvas, final int smd_offset ) {

        for ( int y = 0 ; y < WindowConfig.window_height ; y += smd_offset ) {
            for ( int x = 0 ; x < WindowConfig.window_width ; x += smd_offset ) {
                SMD smd = new SMD( x, y );
                addLevelElement( gameCanvas, smd );
            }
        }
    }

    @Override
    protected boolean addCollectable( final Collectable c ) {
        if ( !collidesWithPlayer( c ) ) {
            if ( !collidesWithLevelElement( c ) ) {
                if ( !collidesWithCollectable( c ) ) {
                    return super.addCollectable( c );
                }
            }
        }
        return false;
    }

    @Override
    protected boolean addLevelElement( final LevelElement levelElement ) {
        if ( !collidesWithPlayer( levelElement ) ) {
            if ( !collidesWithLevelElement( levelElement ) ) {
                return super.addLevelElement( levelElement );
            }
        }
        return false;
    }
}

File: Games/Alex/src/de/hsh/alexander/level/PacManLevel.kt
package de.hsh.alexander.level

import common.actor.Collectable
import common.actor.CollisionCheck.isInBounds
import common.actor.ControllableActor
import common.actor.Level
import common.actor.LevelElement
import common.config.WindowConfig
import common.util.Logger
import de.hsh.alexander.PacManGame
import de.hsh.alexander.actor.ResourcePaths
import de.hsh.alexander.actor.collectables.DataCoin
import de.hsh.alexander.actor.collectables.Invisible
import de.hsh.alexander.actor.level_elements.Condensator
import de.hsh.alexander.actor.level_elements.Fan
import de.hsh.alexander.actor.level_elements.SMD
import de.hsh.alexander.actor.player.PacMan
import de.hsh.alexander.actor.player.PacMan1
import de.hsh.alexander.actor.player.PacMan2
import javafx.beans.property.SimpleIntegerProperty
import javafx.event.EventHandler
import javafx.scene.canvas.Canvas
import javafx.scene.input.KeyEvent
import javafx.scene.input.MouseEvent
import java.util.*

class PacManLevel(gameCanvas: Canvas) : Level(gameCanvas) {

    override fun createLevel(gameCanvas: Canvas) {
        fun createDataCoins(gameCanvas: Canvas) {
            var y = 0.0
            while (y < WindowConfig.window_height) {
                var x = 0.0
                while (x < WindowConfig.window_width) {
                    val d = DataCoin(x, y)
                    if (isInBounds(d, gameCanvas)) add(d)
                    x += 50
                }
                y += 50
            }
        }

        fun addSMDs(gameCanvas: Canvas, smd_offset: Int) {
            var y = 0.0
            while (y < WindowConfig.window_height) {
                var x = 0.0
                while (x < WindowConfig.window_width) {
                    val smd = SMD(x, y)
                    addLevelElement(gameCanvas, smd)
                    x += smd_offset
                }
                y += smd_offset
            }
        }

        fun fillPins(gameCanvas: Canvas) {
            fun addCondensatorsY(gameCanvas: Canvas, start: Int, end: Int, increment: Int, x: Int) {
                var y = start
                while (y < end) {
                    addLevelElement(gameCanvas, Condensator(x.toDouble(), y.toDouble()))
                    addLevelElement(gameCanvas, Condensator((x + increment).toDouble(), y.toDouble()))
                    y += increment
                }
            }

            fun addCondensatorsX(gameCanvas: Canvas, start: Int, end: Int, increment: Int, y: Int) {
                var x = start
                while (x < end) {
                    addLevelElement(gameCanvas, Condensator(x.toDouble(), y.toDouble()))
                    addLevelElement(gameCanvas, Condensator((x + increment).toDouble(), y.toDouble()))
                    x += increment
                }
            }

            addLevelElement(gameCanvas, Condensator(200.0, 600.0, 1))
            var i = 350.0
            while (i < 500) {
                addLevelElement(gameCanvas, Condensator(700.0, i, 0))
                i += 40
            }
            val end = 580
            val increment = 19
            addCondensatorsY(gameCanvas, 285, end, increment, 525)
            addCondensatorsX(gameCanvas, 545, 805, increment, 705)
            addLevelElement(gameCanvas, Condensator(525.0 + increment, (end + increment).toDouble()))
        }

        setBackgroundImage(ResourcePaths.Actor.LevelElements.Backgrounds.microChip,
                background_width,
                background_height)
        addPlayer(PacMan1(250.0, 150.0))
        addPlayer(PacMan2(550.0, 350.0))

        addLevelElement(Fan(200.0, 50.0))
        addLevelElement(Fan(200.0, 375.0, 5.0))
        fillPins(gameCanvas)
        addSMDs(gameCanvas, 200)

        add(Invisible(300.0, 300.0))

        createDataCoins(gameCanvas)
    }

    override fun resetAndThan(gameCanvas: Canvas, thanDo: (Canvas) -> Unit) =
            super.resetAndThan(gameCanvas) {
                gameCanvas.onMouseClicked = EventHandler { clickEvent: MouseEvent ->
                    add(DataCoin(clickEvent.x, clickEvent.y))
                }
                isGameFinished
            }

    val pacMan1Property: SimpleIntegerProperty
        get() = players
                .filterIsInstance<PacMan1>()
                .map { player: PacMan1 -> player.points  }
                .first()

    val pacMan2Property: SimpleIntegerProperty
        get() = players
                .filterIsInstance<PacMan2>()
                .map { player: PacMan2 -> player.points }
                .first()

    override val isGameFinished: Boolean
        get() = if (collectables.size == 0) {
            setChanged()
            this.notifyObservers(PacManGame.gameFinishedMessage)
            true
        } else false

    override infix fun keyboardInput(keyEvent: KeyEvent) =
            players.forEach { pacMan: ControllableActor ->
                pacMan.move(keyEvent)
            }

    override fun update(o: Observable, arg: Any) {
        fun coinCollected(c: Collectable) {
            if (c is DataCoin) {
                val a = c.collector
                if (a is PacMan) a.addPoint()
            }
            collected(c)
        }

        if (o is Collectable) {
            if (arg is String && arg == Collectable.collected) coinCollected(o)
            if (o is Invisible) Logger.log("PacManLevel : invisible collected : $o")
        } else {
            Logger.log(this.javaClass.toString() + ": unknown observable=" + o + " , arg=" + arg)
        }
    }

    override fun add(c: Collectable): Boolean =
            if (!collidesWithPlayer(c) && !collidesWithLevelElement(c) && !collidesWithCollectable(c)) super.add(c)
            else false

    override fun addLevelElement(levelElement: LevelElement): Boolean =
            if (!collidesWithPlayer(levelElement) && !collidesWithLevelElement(levelElement)) super.addLevelElement(levelElement)
            else false

    companion object {
        const val background_width = 950
        const val background_height = 950
    }
}
